Create the Redux Essentials page, take 2

markerikson opened the pull request:
## PR Type

**Does this PR add a _new_ page, or update an _existing_ page?**

Adds the new ~~"Quick Start"~~ **"Redux Essentials"** page

> # Update: Ready!
> 
> Jump ahead to the request for review feedback:
> **https://github.com/reduxjs/redux/pull/3740#issuecomment-650847906**

## Checklist

- [x] Is there an existing issue for this PR?
  - #3674 
- [ ] Have the files been linted and formatted?

## What docs page is being added or updated?

- **Section**: Tutorials
- **Page**: Quick Start

## For Adding New Content

### What kind of content category is this page (tutorial, how-to, explanation, reference)?

The "Quick Start" page is a form of tutorial, but intended to demonstrate the "fastest" / "easiest" / "correct" way to use Redux without explaining _why_ things work this way.  

### Who is the intended target audience?

Beginners who know nothing about Redux, yet.

#### What knowledge are we assuming they have?

- JS syntax: ES6, `async/await`, object spreads
- React: components, hooks
- Other: HTTP / AJAX, `fetch`

### What are the intended results or takeaways from reading this page?

Readers should be able to mimic the demonstrated patterns to create a basic React+Redux app that does some CRUD work.

### What is the most critical info they should learn?

- Redux is a tool for managing state outside of your React components
- The Redux Toolkit package is the simplest and recommended way to use Redux
- Key React+Redux APIs (`configureStore`, `createSlice`, `useSelector` / `useDispatch`)
- Immutability is vital, but either requires hand-written updates or use of RTK+Immer "magic"



## Status

This is a second attempt at writing this new "Quick Start" page, and supersedes #3675 .

Currently an _early_ WIP, but I want the PR preview up so I can start getting feedback as I go.

I've gone ahead and created a new "Tutorials" category, moved the two existing "Basic/Advanced" tutorials under there, and added "Quick Start" under "Tutorials".  This doesn't change any of the existing URLs, it's just sidebar organization.


## Page Link

Direct link to the actual page being worked on:

**https://deploy-preview-3740--redux-docs.netlify.com/tutorials/quick-start/quick-start-part-1**

markerikson commented:
At the moment this is looking like it's going to be two parts:

- Part 1: just explanation of terms and concepts, and walking through the actual structure of a React+Redux app using the standard CRA "counter" template.
- Part 2: hands-on building a small blogging app using RTK.

I've got the first draft content of Part 1 basically done, and will try to put together Part 2 over the next few days.

I'm not convinced this sequencing and explanation is the right approach, but I want to get a complete first draft written so that we have something concrete to get feedback on so we can iterate from there.

markerikson commented:
Been writing a lot of notes and planning for the example app.  Here's a dump of my notes, to give an idea of where I'm trying to go:

## Stuff to show off / talk about

- [x] Creating a slice
    - [x] folder structures
    - [x] same data being used across multiple components
    - [x] adding multiple slices
- [x] `useSelector`
    - [x] basic usage / what is a selector
    - [x] multiple calls in one component
    - [x] careful with `map()` calls and new objects
    - [x] `createSelector`
- [x] local state vs Redux state
    - [x] forms example
- [ ] dispatching actions
    - [x] prepare callback / don't do randomness in reducers
    - [x] multiple slices handling one action
    - [x] same action, multiple places
    - [x] actions as events
- [ ] Reducers
    - [x] only serializable values
    - [x] example of mutating update logic
    - [x] more complex edit logic
    - [ ] normalization
    - [ ] relational updates? (adding a comment to a post)
- [x] some filtering / sorting
    - [x] Sort posts reverse-chronologically (use Reselect, or `createEntityAdapter` sorting?)
    - [x] Filter posts on users page by user
- [x] CRUD API calls
    - [x] Basic data fetching
    - [x] `createAsyncThunk`
    - [x] loading state
    - [x] `createEntityAdapter`
- [ ] time-travel debugging
	- [ ] need at least one example of looking at changes in the store
	- [ ] ideally also one example of fixing something broken somehow


## Outline-ish

### Part 1

- Intro
    - Goals: 
        - Learn the basics of Redux and fastest way to get going, + "the right way"
        - Show enough real-world patterns that you can build an app
    - Prereqs:
        - HTML/CSS, ES6 syntax, React terms, async JS and API calls
- Overview of a Redux App
    - Terms: actions, reducers, store, dispatch, selector, slice?
    - Tools: Redux, RTK, React-Redux
    - Walk through `cra-template-redux` counter example
        - Break down each piece

### Part 2+

- Build a blog app
    - Have a repo / sandbox set up with Mirage, little API client, empty RTK config, basic styling
        - Some light routing
        - Have a "Detailed Explanation" section explaining the basic steps for setup (add RTK and React-Redux, `configureStore`, add `<Provider>`)
    - Initial posts work
        - create posts slice with initial data (concepts: create a slice, folder structure)
        - create `PostsList` and `Post` components
        - render initial list of data (concepts: basic `useSelector` usage)
        - Add post creation page (concepts: dispatching actions; local state vs Redux; forms; mutating update logic)
            - title and text inputs
            - generate UUID and date on creation (concepts: prepare callbacks, serializable values)
        - Add view post page (concepts: same data/multiple components)
        - Add post edit page (concepts: reducer owning state shape)
            - Edit by ID
            - Show some update logic
    - Users and posts
        - Add `usersSlice` and `post.user` (concept: adding multiple slices, basic normalization)
        - Add user dropdown to post creation, and user name to post display (concept: multiple `useSelector` calls in one component)
    - Improving posts
        - Show posts list in reverse-chronological order (concepts: selectors, avoiding mutating data from the store)
            - Try to sort `state.posts`, should get a mutation error
        - Add post reactions (concepts: complex edit logic)
        - Add sidebar with "Recent Posts",  "Popular Posts", and "Total Posts" (concepts: same data / multiple components)
    - Switch to data from the server
        - Fetch posts on mount
            - Show fake API and `client`
            - Fetch data and load in an effect
        - Need to handle loading state
            - Modify `this.timing = 2000` in server for artificial delay
            - Show use of a loading state enum, promise lifecycle actions
    - Add  "Notifications" feature
        - Another slice, more fetching
        - Show unread count as tab badge, mark notifications as read when viewed
        - Add "Mark All as Read" button (concept: same action, multiple places; actions as events)
    - Add "Refresh" button
        - Hit an endpoint that returns 1-3 new posts and 1-5 new notifications (concept: multiple reducers, one action)
        - Similarity to existing "fetch posts" thunk, solved by `createAsyncThunk`
    - Cleanup and optimizations
        - Need to somehow justify showing `createEntityAdapter` and normalized state here
            - Annoying to have to keep doing `someItemsArray.find()` all the time
            - Can normalize, but now how do you show order?  Need selectors
        - Do a React perf capture for a "Mark Notifications Read" click
            - Notifications count badge should render, but why did the posts list render?
            - Use `createSelector` to memoize posts array

markerikson commented:
And fwiw, here's the repo with my **VERY INCOMPLETE** progress so far:

https://github.com/markerikson/redux-quickstart-example-app

The goal atm is to build out the app to an end state I'm happy with in terms of features and code, then restart the process of building it to nail down the intended steps and explanations, and finally turn that into the tutorial content.

Currently, I've got adding and editing posts, a list of posts, fetching users, fetching posts, and showing usernames in posts.

sumitngupta commented:
Awesome work.   Does "Normalization" cover denormalization and denormalization of relational data?

markerikson commented:
@sumitngupta : I'm not planning on going in depth on the topic, if that's what you're asking.  I'm just wanting to show use of `createEntityAdapter` and the ability to look up items as `state.entities[id]`, so that users are aware the API and pattern exist.

leonardoelias commented:
hey, the project is incredible.

can we already start using the react router v6?

sumitngupta commented:
> @sumitngupta : I'm not planning on going in depth on the topic, if that's what you're asking. I'm just wanting to show use of `createEntityAdapter` and the ability to look up items as `state.entities[id]`, so that users are aware the API and pattern exist.

I can appreciate that.  I'm about to spend today/tomorrow working with createEntityAdapter + JSON-API responses + multiple slices + denormalizing the entities together in selectors.  I need to do some research on how to denormalize when the entities state is managed, the current  library assumes the latter format:

`store.getState().users.entities` vs `store.getState().entities.users`.   

Just rambling here, nothing to really respond to, but figured it might be useful to layout my current approach im exploring: 

1. Creating an `entitySlice` which has `createEntityAdapter` adapter associated.  
1. Use that EntitySlice inside all other slices like ArticleSlice, CommentSlice, etc to ensure that all the normalized entities are under the same `entities` state slice. 
1. Ensure (or wrap) the createEntityAdapter's selectors to denormalize objects when used. 

Apologies for probably butchering terminology.  Also sorry if this is not the appropriate place for the thinking. I can delete it.

markerikson commented:
Awright, I think I've completed the example app, as far as what the final implementation is supposed to be.

Here's the repo link again:

https://github.com/markerikson/redux-quickstart-example-app

And a runnable CodeSandbox:

https://codesandbox.io/s/github/markerikson/redux-quickstart-example-app

I'm pretty sure I've hit all the concepts I want to show off somewhere in this codebase, although some of them may be more apparent as intermediate steps along the way.

(I actually wasn't able to get a thrown error regarding mutation from the dev check middleware, because I was trying to sort the array while rendering and Immer has it frozen, so instead you get an error saying you can't overwrite index 0 of the array.)

Now that I know what the final version of the example app ought to look like, I can backtrack and figure out:

- What the base project needs to look like, as users will be cloning / sandbox-forking it when they start the tutorial
- How to approach building this project step by step
- How to teach the concepts as I walk through building it

Like I said, I'm almost definitely trying to cram too much in here.  I'd appreciate any early feedback on the example code.  Also, I have very little design sense, and the current styles are a bashed-together mix of Tania Rascia's styles from her Redux tutorial, the Redux main color, some vaguely-paletted colors I generated from the main color, etc.  I'd certainly appreciate any assistance making this look better.

dpyzo0o commented:
Would there also be a typescript version?

giorgi-m commented:
immer is  a fine idea, but I remember it had some gotchas (https://github.com/immerjs/immer/blob/master/docs/pitfalls.md), not sure how they apply when used in redux toolkit. If they do, might wanna mention in docs.

arjunnaha commented:
Sorry to hijack this thread.

@sumitngupta I would love to chat about your approach with normalisation/de-normalisation, if you managed to get anywhere with it?

OlegLustenko commented on docs/tutorials/quick-start/quick-start-part-3.md:
````diff
@@ -0,0 +1,808 @@
+```jsx title="features/posts/PostsList.js"
+// Sort posts in reverse chronological order by datetime string
+//highlight-start
+const orderedPosts = posts.slice().sort((a, b) => b.date.localeCompare(a.date))
````
Isn't it better to do this calculation in selectors and pass already prepared data to a component?
It seems like we will do all those calculation in run-time, so could be smth like this -
```ts
const selectExcerptPosts = state => {
  const posts = selectPosts(state);
  const orderedPosts = posts
    .slice()
    .sort((a, b) => b.date.localeCompare(a.date));

  return orderedPosts.map(orderedPost => {
    const date = parseISO(timestamp);
    const timePeriod = formatDistanceToNow(date);

    return {
      user: orderedPost.user,
      content: orderedPost.content.substring(0, 100),
      timePeriod,
      timeStamp: orderedPost.postDate
    };
  });
};
```
Later we could apply to reselect here to get the most benefit from selectors

markerikson replied:
Yes and no.

We can't do it in a "plain" selector, as you've showed there, because we're creating a new array reference every time.  If we did this in a selector:

```js
const orderedPosts = useSelector( state => {
  const orderedPosts = state.posts.slice().sort();
  return orderedPosts;
})
```

that would force this component to re-render on _every_ dispatched action, which is not at all what we want.

We _could_ do that in a selector if we were using Reselect.  However, I haven't gotten to that part yet :)  That will come later.

OlegLustenko commented on docs/tutorials/quick-start/quick-start-part-3.md:
```diff
@@ -0,0 +1,808 @@
+}
+
+export const ReactionButtons = ({ post }) => {
+  const reactionButtons = Object.entries(reactionEmoji).map(([name, emoji]) => {
```
It's a matter of style but usually, when I tech somebody React I'm trying to teach them the declarative way of thinking.

```ts
const reactionEmoji = {
  thumbsUp: "👍",
  hooray: "🎉",
  heart: "❤️",
  rocket: "🚀",
  eyes: "👀"
};
const ReactionButton = ({ emoji, children }) => {
  return (
    <button type="button" className="muted-button reaction-button">
      {reactionEmoji[emoji]} {children}
    </button>
  );
};
export const ReactionButtons = ({ post }) => {
  const {
    reactions: { hooray, heart, rocket, eyes, thumbsUp }
  } = post;
  
  return (
    <div>
      <ReactionButton emojiKey="thumbsUp">{thumbsUp}</ReactionButton>
      <ReactionButton emojiKey="hooray">{hooray}</ReactionButton>
      <ReactionButton emojiKey="heart">{heart}</ReactionButton>
      <ReactionButton emojiKey="rocket">{rocket}</ReactionButton>
      <ReactionButton emojiKey="eyes">{eyes}</ReactionButton>
    </div>
  );
};

// later
const onClick = (reaction) => {
  dispatch(reactionAdded({reaction, postId: post.id})
}
```

Just for my curiosity, you have chosen another way to write it, to decrease the number of lines?

markerikson replied:
Yeah.  Either way works, I just didn't see a need to break up the component further.  It's perfectly fine to write larger components - not everything _has_ to be broken down to a tiny level.

markerikson commented on docs/tutorials/quick-start/quick-start-part-3.md:
```diff
@@ -0,0 +1,808 @@
+          payload: {
+            id: nanoid(),
+            // highlight-next-line
+            date: new Date().toISOString(),
```
Note that we're not keeping an actual `Date` in a reducer.  We're creating a new Date instance, and then immediately converting it into a string.

Also, it's not being done in the reducer itself.  As the page says, the "prepare callback" is used to generate the contents of the `action.payload` field.

So yeah, the whole point here is that both the ID generation and the timestamp calculation are being done as the action is created, _before_ it's dispatched, and not in the reducer itself.

nickmccurdy commented:
# Overall
- This article is using normal quotes, but should be using smart quotes.
- It seems like this is going to be significantly longer than the "Basic Tutorial". It feels like this should be renamed to something like "Real World", as it has a different purpose than the Basic tutorial and isn't necessarily quicker besides not going into as much detail about Redux internals.

# Part 1
- Thunks and async stuff in general feel like too much of an advanced topic for part 1. In the same way that Redux's mental model is immutability first, it's also synchronous first. I feel like it's fine to mention this for the first time in part 4.
- The order of introducing immutable state, then createSlice with immer, then immutable state, then immer again is confusing. It's good that users understand both, but Redux is primarily immutable so that should be taught fully before immer (and toolkit's usage of it) are discussed. Maybe just put the entire Rules of Reducers section before this one (Creating Slice Reducers and Actions) since it depends on that anyway.
- Mixing React into all the stuff in the first part feels a bit too complex. I'd prefer to have a separate section right after for React, that way it'll be easier to compare original React state with the React Redux version without unrelated sections in between, and the page could theoretically be swapped out of other frameworks (or even vanilla JS) in the future.

# Part 2
- In the info blurb, createSlice, useSelector, and useDispatch are all mentioned as topics you've learned, but there are already basic explanations on the previous page, so it feels confusing to have both. As I've mentioned under part 1 I think the React stuff belongs in its own section anyway. I'm thinking this could be split into "basic react app" and "real world react app" sections.

# Part 3
- "the component will re-render any time the value returned from useSelector changes to a new reference" this seems to be implying that you need to use memo or useMemo, but useSelector is already memoized
- Users might wonder why the post form state isn't included in Redux, a short explanation about performance and putting global business logic in Redux would be nice.
- `Array.from(array)` and `[...array]` are easier to understand then `array.slice()

markerikson commented:
@nickmccurdy : some thoughts off the top of my head:

- I'm definitely open to renaming it.  Not sure if "Real World" is best, though, as the intent here is that this _is_ something that is meant for true beginners.
- Per the descriptions at the top, the goal here is to show high-level usage patterns that people can imitate, without necessarily understanding all the abstractions that are involved.  This is different than the existing tutorial sequence, which is a low-level, "bottom-up / how it works" approach.
- I'll disagree on the React aspect. The page assumes users are familiar with React already, and since React is part of standard Redux usage, showing them together is reasonable _and_ deliberately intended.
- Similarly, a mention of thunks in part 1 is acceptable given that the context is an overview of the pieces and data flow of a React+Redux app.
- Can look at resequencing the immutability explanations
- `useSelector` is _not_ memoized - that's the point.  
- I don't see `Array.from()` as being easier to understand than `array.slice()`

nickmccurdy commented:
> I'll disagree on the React aspect. The page assumes users are familiar with React already, and since React is part of standard Redux usage, showing them together is reasonable and deliberately intended.
I'm not saying that we should remove React. I agree that it's important, I just feel like the content is different enough that it would be easier to understand if it was introduced together on another page.
`Array.from()` is specifically for creating new Arrays from existing ones that are full length, while `array.slice()` is primarily used to get subsections of arrays immutable`, and users may not be familiar with this behavior

markerikson commented:
I've already got [an entire tutorial series on my blog called "Practical Redux"](https://blog.isquaredsoftware.com/series/practical-redux), so that's out :)

nickmccurdy commented:
I know slice is common, I just disagree with it, I don't think it semantically conveys cloning an array in the way that the others day.

markerikson commented:
# First Draft Done!

It's been _six months_ since [I opened up the issue saying I wanted to write the "Quick Start" page](https://github.com/reduxjs/redux/issues/3674), but I've _finally_ completed a first draft, and it's ready for review.  

(In my defense, a few things happened: the first attempt in #3675 didn't work out, I stopped to do [a ton of research on other tutorials](https://www.notion.so/23c1be94861e41dc82ecb0c2b18aa138?v=d7704772636f4972a4bb066a8ca92bba), RTK 1.3 took up a month, I got assigned to a new project, work's been draining my brain during the week... oh, and a little thing called "COVID-19" threw everything out of whack.)

As a reminder, the PR preview is here:

**https://deploy-preview-3740--redux-docs.netlify.app/tutorials/quick-start/quick-start-part-1**

the example app final source code / commits:

**https://github.com/markerikson/redux-quickstart-example-app/tree/tutorial-steps**

and a CodeSandbox with the final content:

**https://codesandbox.io/s/github/markerikson/redux-quickstart-example-app/tree/tutorial-steps**


## Thoughts

I know this is almost definitely too long.  I just double-checked, and the five pages are 7K, 3K, 5K, 5K, and 6K words respectively.  That's huge!  (And also basically matches my own typical blog post length, so I'm staying true to form, I guess.)

That said, I looked back over [my earlier comment with the list of things I wanted to cover](https://github.com/reduxjs/redux/pull/3740#issuecomment-616253247), and I nailed almost everything except "time-travel debugging" and "multiple reducers, one action".  So, I hit my target, it's just a question of whether I was aiming in the right spot to begin with.


## Requests for Review Focus

I'm specifically looking for review feedback in the following areas:

### Content Structure

- Does the overall layout and sequence of explanations flow logically?
- Do the page divisions and topics work?
- Specifically for the first page:
    - Should we keep the "What is Redux / State Management" sections on this page, to keep it a one-stop explanation, or should we move that content to somewhere under the "Introduction" category?
    - Right now the reader doesn't get into any hands-on coding until page 2.  Is there a way to reorganize things to get them involved sooner?
    - Do the Redux term explanations fit in here okay?
    - Right now the counter example walkthrough goes `store.js` -> `counterSlice.js` -> `Counter.js` -> `index.js`.  Should that be reordered, maybe starting with `index.js` first?
    
### Content Correctness

- Have I actually explained everything right?
- Are the explanations clear, especially for someone who's never used Redux before?
- Am I being too verbose?  Are there ways I can simplify explanations?
- Do the code samples match the explanations?
- I deliberately didn't show every diff, and hand-waved some changes.  Does this style of explanation work okay?
- Do the example repo / CodeSandbox clone and work correctly?
- Is the emphasis on immutability clear enough?

### Editing

- Any typos? Stuff I forgot to finish filling out?
- I know I need to finish filling in some links to Git branches and stuff, that's TBD

### Naming

- I've been calling this the "Quick Start", but it's not exactly "Quick".  Is there a better name?  It _is_ meant for newbies, so the naming needs to reflect that it's meant for beginners.  ("Real World" was suggested, but that feels like it might scare people off.)
- Are there better names / titles for each individual page?
- Right now, the page URLs are `/tutorials/quick-start/quick-start-part-1`.  What should the final URL structure look like?  Maybe `/tutorials/quick-start/part-1-concepts-structure` , etc?

### Other

- How are the "What You'll Learn", "Prerequisites", and "What You Learned" boxes working out?
- Any other "Prerequisites" I should be listing for each page?
- Is using Docusaurus's code block line highlighting working out okay in terms of visibility and emphasis?
- Are the code block file titles sufficient?


I'm really excited to get this finished up and put together, but I fully expect this will need a couple rounds of revision first.

nickmccurdy commented:
# Part 5
- The other pages didn't mention previous pages as prerequisites, so I don't think this section is necessary here.

nickmccurdy commented:
> I've been calling this the "Quick Start", but it's not exactly "Quick". Is there a better name? It is meant for newbies, so the naming needs to reflect that it's meant for beginners. ("Real World" was suggested, but that feels like it might scare people off.)

The words "Tour" and "Overview" are coming to my mind now. I think they explain the goal of teaching the important bits of Redux together better (including other tooling like React Redux and Redux Toolkit). However I still think they're missing something from "practical" and "real world" (I do agree with your reasons for avoiding them but I'm stuck on finding a better alternative).

taniarascia commented on docs/tutorials/quick-start/quick-start-part-1.md:
```diff
@@ -0,0 +1,1107 @@
+
+### Writing Async Logic with Thunks
+
+The next function that's exported from `counterSlice` might look a bit strange:
```
I would add an introduction to this. I would want to know what you mean by "writing async logic with thunks", but it just dives right into some code. Perhaps add a small explanation here, like "JavaScript applications often need to make asynchronous API calls to retrieve data. Redux Thunks are required to do this." or whatever.

valtism commented:
Maybe call it **Core Concepts**

Matvien commented on docs/tutorials/quick-start/quick-start-part-1.md:
```diff
@@ -0,0 +1,1107 @@
+  return async (dispatch, getState) => {
+    try {
+      // make an async call in the thunk
+      const user = userAPI.fetchById(userId)
```
```suggestion
      const user = await userAPI.fetchById(userId)
```

Matvien commented on docs/tutorials/quick-start/quick-start-part-1.md:
````diff
@@ -0,0 +1,1107 @@
+
+Now that we've seen all the different pieces of this application, it's time to circle back to the starting point of this application and see how the last pieces of the puzzle fit together.
+
+```jsx title="src/components/HelloCodeTitle.js"
+function HelloCodeTitle(props) {
+  return <h1>Hello, {props.name}</h1>
+}
+```
````
Maybe I'm wrong, but this looks like a random piece of code pasted here by accident :)

```suggestion
```
```
```

markerikson replied:
Good catch, yeah. I think that's left over from me testing out the new code block title feature from Docusaurus, and I just forgot to remove it.

markerikson commented on docs/tutorials/quick-start/quick-start-part-1.md:
```diff
@@ -0,0 +1,1107 @@
+
+It helps to understand what this "Redux" thing is in the first place. What does it do? What problems does it help me solve? Why would I want to use it?
+
+**Redux is a pattern and library for managing and updating application state, using events called "actions".** It serves as a centralized store for state that needs to be used across your entire application, with rules ensuring that the state can only be updated in a predictable fashion.
```
FWIW, I borrowed that line directly from this article: http://evanxmerz.com/blog/index.php/2020/01/18/redux-explained-in-one-simple-sentence/ .

I do think that the idea of "actions" _is_ central to Redux.  It differentiates Redux from other state management approaches that involve direct updates,  and everything about Redux's design follows from that ("dispatching", "reducer", etc).

But yeah, I'll see if I can reword this.

markerikson commented on docs/tutorials/quick-start/quick-start-part-1.md:
```diff
@@ -0,0 +1,1107 @@
+
+:::warning
+
+**You can _only_ write "mutating" logic in Redux Toolkit's `createSlice` and `createReducer` because they use Immer inside! If you write mutating logic in reducers without Immer, it _will_ mutate the state and cause bugs!**
```
I feel like this part is critical to the explanation sequence. My biggest worry about teaching RTK/Immer as the default is that people will assume you _can_ always mutate state, and then go on to mutate state when they _aren't_ using RTK and promptly cause major bugs. The only solution I have for that is emphasizing, right up front, that you can only "mutate" inside of RTK's APIs.

markerikson commented:
Just pushed a sizeable update:

- Addressed most of the comments from @taniarascia and @Matvien 
  - Removed some unnecessary explanations and a leftover example
  - Clarified several explanations
  - Added a section on "Immutability" under "Concepts", before we even get to "Reducers"
- Split Part 1 by extracting the entire "counter app walkthrough" section into a new Part 2, and renamed the later pages accordingly
- Fixed up several TODO comments

Still debating renaming this.  The best alternate ideas I've had so far are "Essentials Guide" and "Practical Guide".

las3r commented:
### Part 2:
Detailed Explanation: Thunks and Async Logic
`But, in a real Redux app, we're not allowed to import the store into other files, especially in our React components.`
It is allowed and it works, but it is bad practice. Why is this a bad practise? Make us better devs by telling us why ;-)

markerikson commented:
# Ready to Go!

Haven't gotten as much meaningful feedback as I'd hoped, but I think this is basically ready to go.

I've renamed it from "Quick Start" to "Redux Essentials".  It's 25K words and 6 pages, so it's not "quick" at all, but it _is_ the key information you need to get started using Redux.

I added a link to this as "Tutorial" in the header/footer, emphasized it in the "Getting Started" page, and added a new "Tutorial Index" page that describes the differences between the two sets of tutorials.

I also cleaned up the "Getting Started" page to remove most of the external article links, since there were too many of them and people have said they were overwhelming.

I plan on merging this in tomorrow :)

Matvien commented on docs/tutorials/quick-start/quick-start-part-4.md:
```diff
@@ -0,0 +1,808 @@
+  const onAuthorChanged = e => setUserId(e.target.value)
+
+  const onSavePostClicked = () => {
+    if (title && content) {
```
Just a minor refactoring:
```suggestion
    if (canSave) {
```

Matvien commented on docs/tutorials/quick-start/quick-start-part-4.md:
```diff
@@ -0,0 +1,808 @@
+- **Any React component can use data from the Redux store as needed**
+  - Any component can read any data that is in the Redux store
+  - Multiple components can read the same data, even at the same time
+  - Components should try to return just the data they need from selectors
```
I didn't quite get it. 
Does it mean that components should try to extract just the data they need via selectors, not the whole state of the app?

Matvien commented on docs/tutorials/essentials/part-4-using-data.md:
```diff
@@ -0,0 +1,808 @@
+- **Any React component can use data from the Redux store as needed**
+  - Any component can read any data that is in the Redux store
+  - Multiple components can read the same data, even at the same time
+  - Components should try to return just the data they need from selectors
```
I don't get the line 794.
Does it mean that components should try to extract just the data they need via selectors, not the whole state of the app?

markerikson replied:
Exactly.

Matvien commented on docs/tutorials/essentials/part-4-using-data.md:
```diff
@@ -0,0 +1,808 @@
+  const onAuthorChanged = e => setUserId(e.target.value)
+
+  const onSavePostClicked = () => {
+    if (title && content) {
```
Just a minor refactoring:
```suggestion
    if (canSave) {
```

markerikson replied:
You're probably right, but I'm not going to make any more code changes before publishing this because it would require rebasing all the example repo commits as well as the snippet update.

markerikson commented:
Going LIIIIIIVE!
